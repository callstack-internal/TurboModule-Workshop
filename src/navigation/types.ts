import { PerroConEstado } from '../types/dog.ts';

export type RouteNamesParamList = RootStackParamList | DrawerParamList;

export type RootStackParamList = {
  Home: undefined;
  AdoptedDogs: undefined;
  DogDetails: { dog: PerroConEstado };
  HowItWorks: undefined;
};

export type DrawerParamList = {
  Main: undefined;
  HowItWorks: undefined;
};
