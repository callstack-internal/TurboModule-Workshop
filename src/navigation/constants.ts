export const Screens = {
  Main: {
    title: 'Catálogo de Perros',
    name: 'Main',
  },
  HowItWorks: {
    title: 'Cómo funciona',
    name: 'HowItWorks',
  },
  Home: {
    title: 'Home',
    name: 'Home',
  },
  DogDetails: {
    title: 'Detalles',
    name: 'DogDetails',
  },
  Adoptar: {
    title: 'Adoptar',
    name: 'Adoptar',
  },
  MisPerros: {
    title: 'Mis Perros',
    name: 'MisPerros',
  },
} as const;
