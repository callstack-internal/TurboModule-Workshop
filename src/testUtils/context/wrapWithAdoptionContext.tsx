import { AdoptionProvider } from '../../context/AdoptionContext.tsx';

export const wrapWithAdoptionContext = (children: React.ReactNode) => {
  return <AdoptionProvider>{children}</AdoptionProvider>;
};
