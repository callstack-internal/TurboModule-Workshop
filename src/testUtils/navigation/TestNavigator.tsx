import React, { useEffect } from 'react';

import {
  createNavigatorFactory,
  useNavigationBuilder,
  StackRouter,
  TabRouter,
  DrawerRouter,
  DefaultRouterOptions,
  StackRouterOptions,
  TabRouterOptions,
  DrawerRouterOptions,
  NavigationHelpers,
  ParamListBase,
  RouterFactory,
  Router,
  NavigationState,
  NavigationAction,
  NavigationContainer,
} from '@react-navigation/native';
import type {
  StackNavigationState,
  TabNavigationState,
  DrawerNavigationState,
} from '@react-navigation/native';
import type {
  StackActionHelpers,
  TabActionHelpers,
  DrawerActionHelpers,
} from '@react-navigation/native';
import { Text } from 'react-native';

type TestScreenOptions = Record<string, unknown>;

type NavigatorType = 'stack' | 'tabs' | 'drawer';

interface TestNavigatorImplProps extends DefaultRouterOptions {
  type: NavigatorType;
  screenOptions?: TestScreenOptions;
  initialRouteName?: string;
  children: React.ReactNode;
}

type AnyNavigationState =
  | StackNavigationState<ParamListBase>
  | TabNavigationState<ParamListBase>
  | DrawerNavigationState<ParamListBase>;

type AnyRouterOptions = StackRouterOptions | TabRouterOptions | DrawerRouterOptions;

type AnyActionHelpers =
  | StackActionHelpers<ParamListBase>
  | TabActionHelpers<ParamListBase>
  | DrawerActionHelpers<ParamListBase>;

type ExtractedAction = Exclude<
  Parameters<Router<NavigationState, NavigationAction>['getStateForAction']>[0],
  undefined
>;

/**
 * Custom navigator component used for testing React Navigation screens.
 * Supports `stack`, `tabs`, and `drawer` types.
 * Renders all screens at once and injects a `focused` prop based on navigation state.
 *
 * Useful in tests to simulate navigation behavior with a lightweight navigator.
 */
const TestNavigatorImpl = ({ type, ...rest }: TestNavigatorImplProps): React.ReactElement => {
  const router = (
    type === 'drawer' ? DrawerRouter : type === 'stack' ? StackRouter : TabRouter
  ) as RouterFactory<AnyNavigationState, ExtractedAction, AnyRouterOptions>;

  const {
    state,
    descriptors,
    NavigationContent: Content,
  } = useNavigationBuilder<
    AnyNavigationState,
    AnyRouterOptions,
    AnyActionHelpers,
    TestScreenOptions,
    NavigationHelpers<ParamListBase>
  >(router, rest);

  return <Content>{state.routes.map((route) => descriptors[route.key]?.render())}</Content>;
};

/**
 * Creates a navigator factory for the TestNavigator.
 * This wraps the custom `TestNavigator` using `createNavigatorFactory`,
 * enabling use of `Navigator` and `Screen` components just like built-in navigators.
 *
 * Use this when integrating `TestNavigator` into `NavigationContainer`.
 */
const createTestNavigatorFactory = createNavigatorFactory(TestNavigatorImpl);

type LayoutProps = {
  name: string;
  focused: boolean;
  navigation: NavigationHelpers<ParamListBase>;
  children: React.ReactNode;
};

const CustomLayout = ({ name, focused, navigation, children }: LayoutProps): React.ReactElement => {
  useEffect(() => {
    const state = navigation.getState();
    const currentRoute = state.routes[state.index];

    if (focused && currentRoute?.name !== name) {
      navigation.navigate(name);
    } else if (!focused && currentRoute?.name !== 'Dummy') {
      navigation.navigate('Dummy');
    }
  }, [focused, name, navigation]);

  return <>{children}</>;
};

const TestLayout = ({ name, focused, ...props }: LayoutProps): React.ReactElement => {
  return <CustomLayout {...props} name={name} focused={focused} />;
};

export const TestNavigator = ({
  children,
  type = 'stack',
  name = 'TestScreen',
  focused = true,
  screenParams,
  dummyName = 'Dummy',
}: {
  children: React.ReactNode;
  type?: NavigatorType;
  name?: string;
  focused?: boolean;
  screenParams?: object;
  dummyName?: string;
}) => {
  const Test = createTestNavigatorFactory();

  return (
    <NavigationContainer>
      <Test.Navigator
        key={type}
        type={type}
        initialRouteName={focused ? name : dummyName}
        layout={
          /* eslint-disable-next-line react/no-unstable-nested-components */
          (props: Omit<LayoutProps, 'name' | 'focused'>) => (
            <TestLayout {...props} name={name} focused={focused} />
          )
        }
      >
        <Test.Screen name={name} getId={() => name} initialParams={screenParams}>
          {() => children}
        </Test.Screen>
        <Test.Screen name={dummyName} getId={() => dummyName}>
          {() => <Text testID="dummy-screen">Dummy</Text>}
        </Test.Screen>
      </Test.Navigator>
    </NavigationContainer>
  );
};
