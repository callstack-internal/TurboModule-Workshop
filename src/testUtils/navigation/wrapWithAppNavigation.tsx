import React from 'react';

import {
  InitialState,
  NavigationContainer,
  NavigationContainerRef,
} from '@react-navigation/native';
import DrawerNavigator from '../../navigation/navigators/DrawerNavigator.tsx';

type WrapWithAppNavigationProps = {
  initialState?: InitialState;
  navigationRef?: React.RefObject<NavigationContainerRef<any>>;
};

/**
 * Wraps the full app navigator (`DrawerNavigator`) in a `NavigationContainer` for testing.
 *
 * This is the preferred utility when writing tests for **real screen components or components that navigate** that are part of
 * the actual app navigation tree. It gives you a full, realistic navigation environment without mocks,
 * including nested navigators, screen options, transitions, etc.
 *
 * Unlike `TestNavigator`, this uses your actual app's navigator configuration, making it ideal for testing:
 * - Navigation behavior (e.g., `navigate`, `goBack`, deep linking)
 * - Screen lifecycle hooks (e.g., `useFocusEffect`, `useNavigation`)
 * - Integration-level rendering of screen flows
 *
 * ⚠️ Note: This should be used when testing **screen components**.
 * If you're only testing a non-screen component (e.g. custom hook, modal, header)
 * that requires navigation context, consider using `TestNavigator` or `createTestNavigator` instead for a minimal setup.
 *
 */
export const wrapWithAppNavigation = ({
  initialState,
  navigationRef,
}: WrapWithAppNavigationProps) => {
  return (
    <NavigationContainer initialState={initialState} ref={navigationRef}>
      <DrawerNavigator />
    </NavigationContainer>
  );
};
