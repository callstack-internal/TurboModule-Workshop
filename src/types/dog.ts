export interface Perro {
  nombre: string;
  raza: string;
  foto: string;
  edad?: number;
}

export interface Chip {
  chipId: string;
  fechaRegistro: string;
}

export interface NoAdoptable {
  motivo: 'enfermo' | 'muy joven';
  fechaDisponible: string;
}

export type PerroConChip = Perro & Chip;

export type PerroConEstado = Perro | (Perro & NoAdoptable);