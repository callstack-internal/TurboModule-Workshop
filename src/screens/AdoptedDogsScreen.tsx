import React from 'react';
import { View, Text, FlatList, StyleSheet, SafeAreaView, StatusBar } from 'react-native';
import DogCard from '../components/DogCard';
import { useAdoption } from '../context/AdoptionContext';
import { PerroConEstado } from '../types/dog';

function AdoptedDogsScreen() {
  // Get adopted dogs from context
  const { adoptedDogs } = useAdoption();

  // Extract unique key for each item
  const keyExtractor = (item: PerroConEstado, index: number) => `${item.nombre}-${index}`;

  // Render each dog card
  const renderDogCard = ({ item }: { item: PerroConEstado }) => <DogCard perro={item} />;

  // Render empty list component
  const renderEmptyList = () => (
    <View style={styles.emptyContainer}>
      <Text style={styles.emptyText}>No has adoptado ningún perro todavía</Text>
    </View>
  );

  return (
    <SafeAreaView style={styles.container}>
      <StatusBar barStyle="dark-content" backgroundColor="#f8f9fa" />

      <View style={styles.header}>
        <Text style={styles.title}>Mis Perros Adoptados</Text>
      </View>

      <FlatList
        data={adoptedDogs}
        keyExtractor={keyExtractor}
        renderItem={renderDogCard}
        contentContainerStyle={styles.listContent}
        ListEmptyComponent={renderEmptyList}
        showsVerticalScrollIndicator={false}
      />
    </SafeAreaView>
  );
}

const styles = StyleSheet.create({
  container: {
    flex: 1,
    backgroundColor: '#f8f9fa',
  },
  header: {
    padding: 16,
    backgroundColor: '#ffffff',
    borderBottomWidth: 1,
    borderBottomColor: '#e9ecef',
    elevation: 2,
    shadowColor: '#000',
    shadowOffset: { width: 0, height: 2 },
    shadowOpacity: 0.1,
    shadowRadius: 3,
  },
  title: {
    fontSize: 24,
    fontWeight: 'bold',
    marginBottom: 12,
    color: '#212529',
  },
  listContent: {
    padding: 12,
    paddingBottom: 24,
  },
  emptyContainer: {
    padding: 24,
    alignItems: 'center',
    justifyContent: 'center',
    height: 300,
  },
  emptyText: {
    fontSize: 16,
    color: '#6c757d',
    textAlign: 'center',
  },
});

export default AdoptedDogsScreen;
