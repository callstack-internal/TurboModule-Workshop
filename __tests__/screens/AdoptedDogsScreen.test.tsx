import { act, render, waitFor } from '@testing-library/react-native';
import { Screens } from '../../src/navigation/constants.ts';
import { wrapWithAppNavigation } from '../../src/testUtils/navigation/wrapWithAppNavigation.tsx';
import { TestNavigator } from '../../src/testUtils/navigation/TestNavigator.tsx';
import AdoptedDogsScreen from '../../src/screens/AdoptedDogsScreen';
import { PerroConEstado } from '../../src/types/dog';
import AsyncStorage from '@react-native-async-storage/async-storage';
import { wrapWithAdoptionContext } from '../../src/testUtils/context/wrapWithAdoptionContext.tsx';

const initialState = {
  routes: [
    {
      name: Screens.Main.name,
      state: {
        routes: [
          {
            name: Screens.Home.name,
            state: {
              index: 1,
              routes: [{ name: Screens.Adoptar.name }, { name: Screens.MisPerros.name }],
            },
          },
        ],
      },
    },
  ],
};

const component = () => render(wrapWithAdoptionContext(wrapWithAppNavigation({ initialState })));

describe('AdoptedDogsScreen', () => {
  // Sample adopted dogs
  const adoptedDogs: PerroConEstado[] = [
    {
      nombre: 'Max',
      raza: 'labrador',
      foto: 'https://example.com/dog.jpg',
      edad: 3,
    },
    {
      nombre: 'Luna',
      raza: 'golden retriever',
      foto: 'https://example.com/dog2.jpg',
      edad: 2,
    },
  ];

  beforeEach(async () => {
    await AsyncStorage.clear(); // Clear before each test
    jest.clearAllMocks();
  });

  afterAll(async () => {
    await AsyncStorage.clear(); // Clear after all test
  });

  test('renders adopted dogs correctly', async () => {
    await AsyncStorage.setItem('adoptedDogs', JSON.stringify([...adoptedDogs]));

    const { getByText } = component();

    await waitFor(() => {
      // Check if the title is displayed
      expect(getByText('Mis Perros Adoptados')).toBeTruthy();

      // Check if the dogs are displayed
      expect(getByText('Max')).toBeTruthy();
      expect(getByText('Luna')).toBeTruthy();
    });
  });

  test('renders empty state when no dogs are adopted', async () => {
    const { getByText } = component();

    await waitFor(() => {
      // Check if the empty state message is displayed
      expect(getByText('No has adoptado ningún perro todavía')).toBeTruthy();
    });
  });

  test('matches snapshot', async () => {
    const { toJSON } = render(
      <TestNavigator>
        <AdoptedDogsScreen />
      </TestNavigator>,
    );
    await waitFor(() => {
      expect(toJSON()).toMatchSnapshot();
    });
  });
});
