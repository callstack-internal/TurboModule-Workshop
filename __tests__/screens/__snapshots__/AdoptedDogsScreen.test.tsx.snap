// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AdoptedDogsScreen matches snapshot 1`] = `
<RCTSafeAreaView
  style={
    {
      "backgroundColor": "#f8f9fa",
      "flex": 1,
    }
  }
>
  <View
    style={
      {
        "backgroundColor": "#ffffff",
        "borderBottomColor": "#e9ecef",
        "borderBottomWidth": 1,
        "elevation": 2,
        "padding": 16,
        "shadowColor": "#000",
        "shadowOffset": {
          "height": 2,
          "width": 0,
        },
        "shadowOpacity": 0.1,
        "shadowRadius": 3,
      }
    }
  >
    <Text
      style={
        {
          "color": "#212529",
          "fontSize": 24,
          "fontWeight": "bold",
          "marginBottom": 12,
        }
      }
    >
      Mis Perros Adoptados
    </Text>
  </View>
  <RCTScrollView
    ListEmptyComponent={[Function]}
    contentContainerStyle={
      {
        "padding": 12,
        "paddingBottom": 24,
      }
    }
    data={[]}
    getItem={[Function]}
    getItemCount={[Function]}
    keyExtractor={[Function]}
    onContentSizeChange={[Function]}
    onLayout={[Function]}
    onMomentumScrollBegin={[Function]}
    onMomentumScrollEnd={[Function]}
    onScroll={[Function]}
    onScrollBeginDrag={[Function]}
    onScrollEndDrag={[Function]}
    removeClippedSubviews={false}
    renderItem={[Function]}
    scrollEventThrottle={0.0001}
    showsVerticalScrollIndicator={false}
    stickyHeaderIndices={[]}
    viewabilityConfigCallbackPairs={[]}
  >
    <View>
      <View
        style={
          {
            "alignItems": "center",
            "height": 300,
            "justifyContent": "center",
            "padding": 24,
          }
        }
      >
        <Text
          style={
            {
              "color": "#6c757d",
              "fontSize": 16,
              "textAlign": "center",
            }
          }
        >
          No has adoptado ningún perro todavía
        </Text>
      </View>
    </View>
  </RCTScrollView>
</RCTSafeAreaView>
`;
